File: src/LCCI/stacksAndQueues/stack_linked_list.cpp
#include <stdexcept>

template <class T>
class Stack {
private:
    struct StackNode {
        T data;
        StackNode* next;
    };
    StackNode* top;

public:
    Stack() : top(nullptr) {}

    ~Stack() {
        while (!isEmpty()) pop();
    }

    void push(const T& item) {
        StackNode* node = new StackNode();
        node->data = item;
        node->next = top;
        top = node;
    }

    T pop() {
        if (top == nullptr) throw std::out_of_range("pop from an empty stack");
        T item = top->data;
        StackNode* temp = top;
        top = top->next;
        delete temp;
        return item;
    }

    T peek() const {
        if (top == nullptr) throw std::out_of_range("peek from an empty stack");
        return top->data;
    }

    bool isEmpty() const {
        return top == nullptr;
    }
};

File: src/LCCI/stacksAndQueues/stack_linked_list.py
"""
stack LIFO implementaton using a singly linked list in python

"""
class StackNode:
    def __init__(self , data):
        self.data = data
        self.next = None

class Stack:
    def __init__(self):
        self.top = None

    def push(self, item):
        node = StackNode(item)
        node.next = self.top
        self.top = node

    def pop(self):
        if self.top is None:
            raise IndexError("stack is empty")
        item = self.top.data
        self.top = self.top.next
        return item

    def peek(self):
        if self.top is None:
            raise IndexError("stack is empty")
        return self.top.data

    def is_empty(self):
        return True if self.top is None else False


"""
Python's built-in list already is a dynamic array with O(1) amortized append/pop from the end, 
so the idiomatic, everyday Python stack is just:

stack = []
stack.append(10)   # push
stack.append(20)
stack.pop()         # pop -> 20
stack[-1]           # peek -> 10
len(stack) == 0     # is_empty

"""